src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -134,7 +134,31 @@ public SwerveModuleState[] convertToModuleStates(double xTranslation, double yTr
         }
         SmartDashboard.putString("chassis speeds",chassisSpeeds.toString());
         //Convert Chassis Speeds to individual module states
-        return kDriveKinematics.toSwerveModuleStates(chassisSpeeds);  
+        return kDriveKinematics.toSwerveModuleStates(chassisSpeeds);
+    }
+
+    private void normalizeDrive(SwerveModuleState[] desiredStates, ChassisSpeeds speeds){
+        //Find magnitude of translation input, map to a scale of 1 in order to be comparable to rotation
+        double translationalK = Math.hypot(speeds.vxMetersPerSecond, speeds.vyMetersPerSecond) / kPhysicalMaxSpeed;
+        //Find magnitude of rotation input, map to a scale of 1 in order to be comparable to translation
+        double rotationalK = Math.abs(speeds.omegaRadiansPerSecond) / kPhysicalMaxAngularSpeed;
+        //Use the larger of the two magnitudes for scaling
+        double k = Math.max(translationalK, rotationalK);
+      
+        //Find the how fast the fastest spinning drive motor is spinning                                       
+        double realMaxSpeed = 0;
+        for (SwerveModuleState moduleState : desiredStates) {
+          realMaxSpeed = Math.max(realMaxSpeed, Math.abs(moduleState.speedMetersPerSecond));
+        }
+        
+        if(realMaxSpeed != 0){
+            //Map input magnitude back to speed in meters per second, divide by real speed to create scale
+            double scale = Math.min(k * kPhysicalMaxSpeed / realMaxSpeed, 1);
+            //Desaturate speeds using that scale
+            for (SwerveModuleState moduleState : desiredStates) {
+              moduleState.speedMetersPerSecond *= scale;
+            }
+        }
```

Comment:
This is stale, we removed this from the swerve code because it didn't do what we wanted.